### PR TITLE
initial implementation of DANN

### LINKED ISSUE #14: implementing domain adversarial neural network (DANN)
Now that we've demonstrated that the NPEs are not robust across simulations, lets implement some domain adaptation methods. In particular, this issue will track the implementation of DANN (https://github.com/fungtion/DANN?tab=readme-ov-file)

### PR DESCRIPTION
**Pull request by Nikhil0504**
This pull request introduces significant new functionality to the `haloflow` package, specifically adding a Domain Adversarial Neural Network (DANN) model for domain adaptation tasks. The changes include the addition of new classes and methods for data loading, model definition, training, evaluation, and visualization.

#### Data Loading:
* [`src/haloflow/dann/data_loader.py`](diffhunk://#diff-e8c2ecdbc03f5c11be5e1ac81af1ee6838c57c3513c07611790518751e04832eR1-R69): Added `SimulationDataset` class to handle loading and preprocessing of simulation data, and to provide DataLoaders for training and testing.

#### Model Definition:
* [`src/haloflow/dann/model.py`](diffhunk://#diff-2c61fb01f7cbd69e9696314cf7a7bff3d4f23cf2fdf0702b8189323bc3dbc535R1-R42): Added `DANN` class for the Domain Adversarial Neural Network model, including feature extractor, label predictor, and domain classifier components.

#### Training:
* [`src/haloflow/dann/train.py`](diffhunk://#diff-87e74af9a905bc8f8336820d6ad4d33cb1289a7cb94ded749beb31260d7c2cffR1-R67): Added `train_dann` function to train the DANN model, incorporating both task-specific and domain classification losses.

#### Evaluation:
* [`src/haloflow/dann/evalutate.py`](diffhunk://#diff-97a71e154854b84646b7a5cd4f740413f8494e331793383f374b9510fdc2f6b1R1-R75): Added `evaluate_regression` and `domain_accuracy` functions to evaluate the model's regression performance and domain classification accuracy, respectively.

#### Utility:
* [`src/haloflow/dann/utils.py`](diffhunk://#diff-5c0c182b75e73fdc77f63d6d5189852698a7535b526e12943e741d8dca808aa5R1-R21): Added `GradientReversal` and `GradientReversalLayer` classes to implement gradient reversal for adversarial training.

#### Visualization:
* [`src/haloflow/dann/visualise.py`](diffhunk://#diff-13d4e11b31af7d22e69d006e53e194c856001be15c4c110571165ea762f13d0fR1-R104): Added functions to visualize features using t-SNE, including `visualize_features_fast` and `plot_combined_tsne`.
image

### CODE REVIEW

File: src/haloflow/dann/model.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class DANN(nn.Module):
+    def __init__(self, input_dim, num_domains=4, alpha=1.0):
+        super().__init__()
+        self.alpha = alpha
+
+        # Feature Extractor (Shared)
+        self.feature_extractor = nn.Sequential(
+            nn.Linear(input_dim, 128), nn.ReLU(), nn.Linear(128, 64), nn.ReLU()
```
Comment (Nikhil0504) [inline src/haloflow/dann/model.py]: Is it okay to set 128 and then to 64 here only with 2 layers? Or do we need to add more layers?
@changhoonhahn

Comment (changhoonhahn) [reply src/haloflow/dann/model.py]: I think that's good enough to start. However, I would implement the DANN object so that you can specify the architecture when you initialize it: something along the lines of
```python 
class DANN(nn.Module):
    def __init__(self, input_dim, layers=[128, 64], num_domains=4, alpha=1.0):
        super().__init__()
        self.alpha = alpha

        # Feature Extractor (Shared)
        _layers = [nn.Linear(input_dim, layers[0]), nn.ReLU()] 
        for i in range(len(layers)): 
            _layers.append(nn.Linear(layers[i], layers[i+1]))
            _layers.append(nn.Relu())

        self.feature_extractor = nn.Sequential(*_layers)
```

That way you can adjust the layers later. It's likely that with too shallow of a NN the compression feature extraction will be very lossy (i.e. M* and Mh constraints will be broad) so we may have to adjust that.

**Comment by changhoonhahn** [comment]
Do you understand why Domain 0 has that region with no overlap?

Does Domain 0 have a higher mass resolution? I'm wondering if those blue points are low mass galaxies.

**Comment by Nikhil0504** [comment]
> Do you understand why Domain 0 has that region with no overlap?

Yes, I think so.

> Does Domain 0 have a higher mass resolution? I'm wondering if those blue points are low mass galaxies.

Yeah, Domain 0 is `TNG 50`. `TNG 50` should have a higher resolution than all the other simulations. The test simulation (Domain 4/ Test Domain) is `TNG_ALL`.

**Comment by changhoonhahn** [comment]
I see. For DA purposes, we should impose a consistent mass limit across the simulations. DA should not work for the low mass TNG50 galaxies, since they lie below the mass limits of the other simulations, so it's a waste of training.

File: src/haloflow/dann/data_loader.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        self.sims = sims
+        self.obs = obs
+        self.data_dir = data_dir
+        self.data = self._load_data()
+
+    def _load_data(self):
+        data = {}
+        for sim in self.sims:
+            Y_train, X_train = D.hf2_centrals("train", self.obs, sim=sim)
+            Y_test, X_test = D.hf2_centrals("test", self.obs, sim=sim)
```
Comment (Nikhil0504) [inline src/haloflow/dann/data_loader.py:20]: So, should we impose the mass priors here?

```python
# impose mass priors (already in log space)
            mass_range_sm = ...
            mass_range_hm = ...
            mask_sm = (Y_train[:, 0] > mass_range_sm[0]) & (Y_train[:, 0] < mass_range_sm[1])
            mask_hm = (Y_train[:, 1] > mass_range_hm[0]) & (Y_train[:, 1] < mass_range_hm[1])
            Y_train = Y_train[mask_sm & mask_hm]
```

@changhoonhahn, could you tell what mass ranges we should have for stellar and halo mass?

Comment (changhoonhahn) [reply src/haloflow/dann/data_loader.py:20]: For now, I would identify the stellar mass and halo mass limits for all of the simulations. Then use the highest limit as the universal one. 

In the future, though, we'll need to think about it. There's the mass limit that corresponds to the literal minimum mass. There's also the limit that corresponds to when the sample is no longer complete.

**Comment by Nikhil0504** [comment]
A lot better with the new priors:
image

See:
https://github.com/galactic-ai/haloflow/pull/16/commits/8550cfba8b625b20c7d10cad0a055d960d65b3cf#diff-e8c2ecdbc03f5c11be5e1ac81af1ee6838c57c3513c07611790518751e04832eL22-R29

**Comment by changhoonhahn** [comment]
Looked through the pull request. Everything looks good, although some of the modules may be a bit excessive, haha.
